framework/test/util/TestUtils.cpp

```diff
@@ -0,0 +1,490 @@
+#else
+  char* tempdir = getenv("TMPDIR");
+  return std::string(((tempdir == nullptr)?"/tmp":tempdir));
+#endif
+}
+
+static const char validLetters[] = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+
+// A cross-platform version of the mkstemps function
+static int mkstemps_compat(char *tmpl, int suffixlen)
+{
+  static unsigned long long value = 0;
+  int savedErrno = errno;
+
+// Lower bound on the number of temporary files to attempt to generate.
+#define ATTEMPTS_MIN (62 * 62 * 62)
+
+/* The number of times to attempt to generate a temporary file.  To
+   conform to POSIX, this must be no smaller than TMP_MAX.  */
+#if ATTEMPTS_MIN < TMP_MAX
+  const unsigned int attempts = TMP_MAX;
+#else
+  const unsigned int attempts = ATTEMPTS_MIN;
+#endif
+
+  const std::size_t len = strlen(tmpl);
+  if ((len - suffixlen) < 6 || strncmp(&tmpl[len - 6 - suffixlen], "XXXXXX", 6))
+  {
+    errno = EINVAL;
+    return -1;
+  }
+
+  /* This is where the Xs start.  */
+  char *XXXXXX = &tmpl[len - 6 - suffixlen];
+
+/* Get some more or less random data.  */
+#ifdef US_PLATFORM_WINDOWS
+  {
+    SYSTEMTIME stNow;
+    FILETIME ftNow;
+
+    // get system time
+    GetSystemTime(&stNow);
+    stNow.wMilliseconds = 500;
+    if (!SystemTimeToFileTime(&stNow, &ftNow))
+    {
+      errno = -1;
+      return -1;
+    }
+    unsigned long long randomTimeBits = ((static_cast<unsigned long long>(ftNow.dwHighDateTime) << 32) |
+                                         static_cast<unsigned long long>(ftNow.dwLowDateTime));
+    value = randomTimeBits ^ static_cast<unsigned long long>(GetCurrentThreadId());
+  }
+#else
+  {
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+    unsigned long long randomTimeBits =
+      ((static_cast<unsigned long long>(tv.tv_usec) << 32) | static_cast<unsigned long long>(tv.tv_sec));
+    value = randomTimeBits ^ static_cast<unsigned long long>(getpid());
+  }
+#endif
+
+  for (unsigned int count = 0; count < attempts; value += 7777, ++count)
+  {
+    unsigned long long v = value;
+
+    /* Fill in the random bits.  */
+    XXXXXX[0] = validLetters[v % 62];
+    v /= 62;
+    XXXXXX[1] = validLetters[v % 62];
+    v /= 62;
+    XXXXXX[2] = validLetters[v % 62];
+    v /= 62;
+    XXXXXX[3] = validLetters[v % 62];
+    v /= 62;
+    XXXXXX[4] = validLetters[v % 62];
+    v /= 62;
+    XXXXXX[5] = validLetters[v % 62];
+
+#ifdef US_PLATFORM_WINDOWS
+    int fd = _open(tmpl, O_RDWR | O_CREAT | O_EXCL, _S_IREAD | _S_IWRITE);
+#else
+    int fd = open(tmpl, O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
+#endif
+    if (fd >= 0)
+    {
+      errno = savedErrno;
+      return fd;
+    }
+    else if (errno != EEXIST)
+    {
+      return -1;
+    }
+  }
+
+  /* We got out of the loop because we ran out of combinations to try.  */
+  errno = EEXIST;
+  return -1;
+}
+
+// A cross-platform version of the POSIX mkdtemp function
+char* mkdtemps_compat(char *tmpl, int suffixlen)
```

Comment:
It looks like there's a lot of common code between mkdtemps_compat and mkstemps_compat that can be factored out to remove the duplication?

Reply:
Mostly for the same reason as above. Laziness, working on bugs or features, afraid of introducing regressions, ...